### PR TITLE
distsql: de-duplicate equivalent local aggregations

### PR DESCRIPTION
**Pull request by richardwu**
Closes https://github.com/cockroachdb/cockroach/issues/18534

## Example
For the following contrived example
```
SELECT sum(a), avg(a), count(a), stddev(a), variance(a) FROM data
```
we de-duplicate __a ton__ of local aggregations
image
where the [current state is in the left image](https://cockroachdb.github.io/distsqlplan/decode.html?eJzUlVGLGjEQx9_7KZZ5sjBQs7t63j5FzhMWWm1X717KcqRmWARvI0kWWg6_e4lb8BRNyvkg-5bJ5J9f5p-BeYNaSZqJVzKQ_QQGCDEgJICQAsIASoStVisyRml3pBXk8jdkfYR1vW2s2y4RVkoTZG9g13ZDkMFS_NpQQUKSBgRJVqw3e8hWr1-F_sOlsAIQ5o3NIs6g3CGoxh4uNFZUBBnb4f9Dx1WlqRJWnTAXT996nH0GfLd6mD_NlmfWix_FJJ9OvefDZy6VE18s51BFUystSZM8KqLcfbzg-N_qJXfPS46i1EXTfDb--rJYTiaPzz0-QD5EfndIPI-LfDx7eOzxEfJ75KzvcgXVkrT7PYx4_IUnGPEUIz7AiA8vOpAcOcBu0UUBaNe6KL6FhwFo1zxMbuFhANo1D9NbeBiAds3DwIAryGxVbehkMpy_ue8mBsmK2vFiVKNX9F2r1R7ThvO9br8hydg2y9ogr9uUe-B7MfOK4yMxOxXHfnIAnXjVqV-cXvPugVc89JOH15DvvOKRnzy6hnzv_6t-oE38TXbKLnef_gYAAP__8k43jg==) and the [__improvement is in the right__](https://cockroachdb.github.io/distsqlplan/decode.html?eJzElFFr2zAQx9_3Kcw9dXCwyHayzk8KTQOGLdmctC_DFC06jCG1jCTDRvF3H7YHrb1GHgup33w6__W7_524JyiUpI14JAPRd2CA4ANCAAghIMwhRSi1OpAxSje_dIJY_oRohpAXZWWb4xThoDRB9AQ2t0eCCPbix5ESEpI0IEiyIj-2kFLnj0L_4lJYAQjbykYeZ5DWCKqyzxcaKzKCiNX479BllmnKhFUD5u7uyxVn7wHhZnu32f_53n1LVvF63Uan8P5J_DO1KpSWpEn2oGn9XwX2vh7iplj_72gdb5afH3b71er2_ooHyBm-TNwvk3i5ubntpRIqJOmm2ehx_wMP0OMhenyOHl-cbEDQawCbYugj0EsP3Z_C8wj00p6DKTyPQC_tOZzC8wj0LRfaK_iETKkKQ4PF9vrNs2bhkcyo245GVfpAX7U6tJgu3La69kCSsV2WdUFcdKmmwJdi5hT7PTEbin03eQQdONWhWxyeU_fcKV64yYtzyB-d4ms3-foc8if3rGYjz8T9yIbstH73OwAA__-7q92w). 

## Follow-up

We can do the same thing with final aggregations as well (before the rendering stage). 

cc: @RaduBerinde @vivekmenezes

### CODE REVIEW

File: pkg/sql/distsqlrun/processors.go [hunk 1]
```diff
@@ -500,7 +500,7 @@ func (a AggregatorSpec_Aggregation) Equals(b AggregatorSpec_Aggregation) bool {
 			return false
 		}
 	} else {
-		if a.FilterColIdx == nil || *a.FilterColIdx != *b.FilterColIdx {
+		if b.FilterColIdx == nil || *a.FilterColIdx != *b.FilterColIdx {
```
Comment (richardwu) [inline pkg/sql/distsqlrun/processors.go]: This was a neat little bug that surfaced.

**Comment by richardwu** [comment]
As a point of reference, query 1 of TPC-H (scale factor 1, 3-node local cluster) which duplicates 2 local aggregations (out of 13 local aggregations) took 63.71 seconds with this branch vs 97.48 seconds on master.

File: pkg/sql/distsqlrun/processors.go (hunk 1 above)
Comment (RaduBerinde) [reply pkg/sql/distsqlrun/processors.go]: Consider putting this and any associated test in a separate commit (would be easier to look at if someone looks at code history)

**Comment by RaduBerinde** [comment]
LGTM

**Comment by richardwu** [comment]
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/distsqlrun/processors.go, line 503 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18810#-Kv-FFJ5tY_r2tbCnH5H-r1-503:-Kv9iUMruZ1s0dyGQjIP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/010128fb374e7755b3e3cbcd9d8ef7ebefdc6554/pkg/sql/distsqlrun/processors.go#L503)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Consider putting this and any associated test in a separate commit (would be easier to look at if someone looks at code history)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18810)*

**Pull request merged**